cluster-controller/src/main/java/io/strimzi/controller/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -0,0 +1,250 @@
+        // TODO Should we annotate the PVCs?
+        boolean deleteClaims = zk != null && zk.getStorage().type() == Storage.StorageType.PERSISTENT_CLAIM
+                && zk.getStorage().isDeleteClaim();
+        List<Future> result = new ArrayList<>(4 + (deleteClaims ? zk.getReplicas() : 0));
+
+        result.add(configMapOperations.reconcile(namespace, ZookeeperCluster.zookeeperMetricsName(name), null));
+        result.add(serviceOperations.reconcile(namespace, ZookeeperCluster.zookeeperClusterName(name), null));
+        result.add(serviceOperations.reconcile(namespace, ZookeeperCluster.zookeeperHeadlessName(name), null));
+        result.add(zkSetOperations.reconcile(namespace, ZookeeperCluster.zookeeperClusterName(name), null));
+
+        if (deleteClaims) {
+            for (int i = 0; i < zk.getReplicas(); i++) {
+                result.add(pvcOperations.reconcile(namespace, zk.getPersistentVolumeClaimName(i), null));
+            }
+        }
+
+        // TODO wait for endpoints and pods to disappear
+
+        return CompositeFuture.join(result);
+    };
+
+    private final Future<ReconcileResult<Void>> createOrUpdateTopicController(Reconciliation reconciliation, ConfigMap assemblyCm) {
+        String namespace = assemblyCm.getMetadata().getNamespace();
+        String name = assemblyCm.getMetadata().getName();
+        log.info("{}: create/update topic controller {}", reconciliation, name);
+        TopicController topicController = TopicController.fromConfigMap(assemblyCm);
+        Deployment deployment = topicController != null ? topicController.generateDeployment() : null;
+        return deploymentOperations.reconcile(namespace, topicControllerName(name), deployment);
+    };
+
+    private final Future<ReconcileResult<Void>> deleteTopicController(Reconciliation reconciliation) {
+        String namespace = reconciliation.namespace();
+        String name = reconciliation.assemblyName();
+        log.info("{}: delete topic controller {}", reconciliation, name);
+        return deploymentOperations.reconcile(namespace, topicControllerName(name), null);
+        // TODO wait for pod to disappear
+    };
+
+    @Override
+    protected void delete(Reconciliation reconciliation, Handler<AsyncResult<Void>> handler) {
+        Future<Void> f = Future.<Void>future().setHandler(handler);
+        deleteTopicController(reconciliation)
+                .compose(i -> deleteKafka(reconciliation))
+                .compose(i -> deleteZk(reconciliation))
+                .compose(ar -> f.complete(), f);
+    }
+
+    @Override
+    protected List<HasMetadata> getResources(String namespace, Labels selector) {
+        List<HasMetadata> result = new ArrayList<>();
+        result.addAll(kafkaSetOperations.list(namespace, Labels.forType(KafkaCluster.TYPE)));
+        result.addAll(zkSetOperations.list(namespace, Labels.forType(KafkaCluster.TYPE)));
```

Comment:
Do the Zookeeper resources really have as a type label `KafkaCluster.TYPE`? Shouldn't these be:
```
result.addAll(zkSetOperations.list(namespace, Labels.forType(ZookeeperCluster.TYPE)));
result.addAll(deploymentOperations.list(namespace, Labels.forType(TopicController.TYPE)));
result.addAll(serviceOperations.list(namespace, Labels.forType(KafkaCluster.TYPE)));
result.addAll(serviceOperations.list(namespace, Labels.forType(ZookeeperCluster.TYPE)));
```

Reply:
This is a really good point. Before I think we needed a label for the different clusters (kafka, zk), but I don't think we do now:

* For createOrUpdate() we identifiy resources by name, not by label.
* We still use labels when doing a periodic reconciliation (to identify what assemblies exist but which now lack a CM), but in this case we don't need to make a ZK/Kafka distinction: We just need to figure out the assembly name for which there's no matching CM. This works so long as the labels of all the resource share a common `type`, and each have their respective `cluster`: We use the `type` to find the resources and them map to the name of the cluster/assembly CM which owns them via the `cluster`.

So the CC doesn't really consume these zookeeper or topic controller labels itself, currently. I admit this has come about as a result of the refactoring, rather than being a conscious decision.

As I mentioned in a comment to @ppatierno ealier I think we should tidy this up. The `Labels.*type*()` methods should use something other than a `String` (probably an `enum`), and the `*cluster*()` methods should use an `enum`. This will make the API less confusing and more type-safe. We should also consider renaming the label `cluster` → `assembly`. I also think maybe `type` →`assembly.type` would also be more clear, though it's less concise. 

Thoughts?

Reply:
Wait ... what happens if I had a running Kafka cluster and now I delete the config map together with all Kafka resources (Kafka SS, Kafka services), but leave all Zookeeper resources? Will the reconciliation find them and delete them? With the current code it would simply ignore them, or?

Reply:
I have now changed the code so that the reconciliation would find and delete the zk resources.